p4proto/kctrl/switchapi/README.md

```diff
@@ -0,0 +1,165 @@
+ * **switch_api_neighbor_delete**: Retrieve the neighbor_info from neighbor handle and call the platform dependent (PD) API's to delete table entry from target backend and delete the handle
+
+## Nexthop <a name="nexthop"></a>
+ * **switch_api_nhop_create**: Handle is allocated and initialized for handle type SWITCH_HANDLE_TYPE_NHOP. Check if the nhop handle exists, if not, create one and get the nhop key based on nhop type. Once the handle is successfully created, cache the nhop info with key as nhop key in the hashtable and return nhop handle to SwitchSAI
+ * *Note*:
+Backend programming for nhop create is not needed because neighbor_create will call a nexthop_create always before creating neighbor
+ * **switch_api_nhop_delete**: Retrieve the nhop_info from nhop handle and if nhop reference count is more than 1, then decrement the count and return. Else, delete the cached information from hastable for this nhop key.
+
+## Route <a name="route"></a>
+ * During initialization, device context is set, route hashtable is initialized and handle is allocated for  SWITCH_HANDLE_TYPE_ROUTE and updated in device context
+ * **switch_api_l3_route_add**:
+    * Create route entry using vrf handle and ip address from api_route_info attribute passed by SwitchSAI and lookup the hashtable to search if this route entry already exists. If yes, then return to the SwitchSAI else, create the handle of type SWITCH_HANDLE_TYPE_ROUTE.
+    *  Once the handle is successfully created, if nhop handle is of type SWITCH_HANDLE_TYPE_NHOP,  call the platform dependent(PD) API to add entry to ipv4 table in target backend. If nhop handle is of type SWITCH_HANDLE_TYPE_ECMP_GROUP, call the platform dependent(PD) API to add entry to ip4 table  and ecmp_hash table
+    * Once the table entry is successfully added to backend, cache the route_info in the hastable with route entry as the key
+
+ * **switch_api_l3_route_delete**: Retrieve the route handle from hashtable using route entry as the key. Call the platform dependent(PD) API to delete the table entries from target backend and delete the route handle.
+
+## Platform Dependent(PD) API <a name="platform_dependent_api"></a>
+
+All platform dependent (PD) API's allocates a session handle for TDI transactions and retrieves the key and table handle and id's associated for action and all required fields with appropriate TDI API's and program the target backend using tdi_table_entry_add API. Once the table is successfully programed, the session handle is deleted
+
+#### FDB platform dependent API:
+ * **switch_pd_l2_tx_forward_table_entry**
+   *  For FDB learns from vlan interface
+```
+    Table name: LNW_L2_FWD_TX_TABLE
+    Key: LNW_L2_FWD_TX_TABLE_KEY_DST_MAC
+    Action: LNW_L2_FWD_TX_TABLE_ACTION_L2_FWD with parameters port id of vlan interface
+```
+   * For FDB learns from physical interface
```

Comment:
these doesnt show as sub sections in, it shows in the same level as switch_pd_l2_tx_forward_table_entry
https://github.com/ipdk-io/ovs/blob/linux_networking_documentation/p4proto/kctrl/switchapi/README.md#fdb-platform-dependent-api